File: kattis/catalansquare/catalan1.cs
using System;
using System.Numerics;

public static class Program
{
    static BigInteger Product(Int32 x, Int32 dx, Int32 mx)
    {
        var p = BigInteger.One;
        for (Int32 i = x; i <= mx; i += dx)
            p *= i;
        return p;
    }
    
    static BigInteger Catalan(Int32 n)
    {
        Int32 c = n / 2;
        Int32 d = n % 2;
        Int32 m = n - c;
        
        Int32 g = 0;
        for (Int32 i = m; (i /= 2) > 0; )
            g += i;
        
        Int32 k = 1 << (c - g);
        BigInteger top = k * Product(n - d + 3, 2, n * 2);
        
        BigInteger bot = 1;
        for (Int32 i = m; i > 0; i /= 2)
            bot *= Product(1, 2, i);
        
        return top / bot;
    }
    
    public static void Main()
    {
        var ln = Console.ReadLine();
        var n = Int32.Parse(ln);
        
        var c = Catalan(n + 1);
        
        Console.WriteLine(c);
    }
}

File: kattis/catalansquare/catalan1.java
import java.util.Scanner;
import java.math.BigInteger;

public class catalan
{
    static BigInteger product(int xo, int dx, int mx)
    {
        BigInteger p = BigInteger.ONE;
        for (int x = xo; x <= mx; x += dx)
            p = p.multiply(BigInteger.valueOf(x));
        return p;
    }
    
    static BigInteger catalan(int n)
    {
        int c = n / 2;
        int d = n % 2;
        int m = n - c;
        
        int g = 0;
        for (int i = m; (i /= 2) > 0; )
            g += i;
        
        BigInteger k = BigInteger.valueOf(1 << (c - g));
        BigInteger top = product(n - d + 3, 2, n * 2).multiply(k);
        
        BigInteger bot = BigInteger.ONE;
        for (int i = m; i > 0; i /= 2)
            bot = bot.multiply(product(1, 2, i));
        
        return top.divide(bot);
    }
    
    public static void main(String[] args)
    {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        
        BigInteger c = catalan(n + 1);
        
        System.out.println(c);
    }
}